source/fundamentals/typescript.txt

```diff
@@ -142,6 +142,127 @@ section.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+How you specify the ``_id`` field in type parameters passed to your
+``Collection`` instance affects the behavior
+of insert operations. The following table describes how different
+``_id`` field specifications affect insert operations:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 20 20 20 40
+
+   * - ``_id`` field type
+     - Example Type
+     - Required on insert
+     - Behavior on insert
+
+   * - | Unspecified
+     - | Not applicable
+     - | No
+     - | The driver creates an 
+         :manual:`ObjectId </reference/method/ObjectId/>`
+         value for each inserted document.
+
+   * - | Specified
+     - | ``{ _id: number };`` 
+     - | Yes
+     - | If you do not specify a value for the ``_id`` field in an insert operation,
+         the driver raises an error.
+
+   * - | Specified as optional
+     - | ``{ _id?: number };``
+     - | No
+     - | If you do not specify the ``_id`` field in an insert operation,
+         the driver adds an ``_id`` field value generated by a
+         **primary key factory**.
+       | A primary key factory is a class that defines a system for creating
+         unique identifiers for documents and implements the ``PkFactory``
+         interface.
+       | The default primary key factory generates ``_id`` values of type
+         ``ObjectId``.
+
+.. important:: Data Consistency and Optional _id Types
+
+   If you do not specify a primary key factory method that generates
+   ``_id`` values of your desired type, you may experience data inconsistency.
+   This occurs because the driver's default primary factory method generates
+   ``_id`` values of type ``ObjectId`` regardless of the type you specify for the 
+   ``_id`` field in your type parameter.
+
+   To show this behavior, assume you apply the following interface to
+   your ``Collection`` instance:
+
+   .. code-block:: typescript
+
+      interface I {
+        _id: number;
+        name: string;
+      } 
+
+   The following sequence of insert operations inserts one document
+   with an ``_id`` of type ``number`` and one document with an 
+   ``_id`` of type ``ObjectId`` into your collection:
+
+   .. code-block:: typescript
+
+      collection.insertOne({ _id: 1, name: "Scar" });
+      collection.insertOne({ name: "Ursula" });
+
+   To ensure that your data is consistent, apply a primary key factory to
+   your ``MongoClient`` instance like this:
+
+   .. code-block:: typescript
+
+      const client = new MongoClient(uri, {
+        pkFactory: { createPk: () => Math.floor(Math.random() * Number.MAX_SAFE_INTEGER) },
+      });
+
+If you must specify the ``_id`` field in the type you define to represent
+documents in your collection but you do not want to specify values for the
+``_id`` field in insert operations, use the ``OptionalId`` type. The
+``OptionalId`` type accepts a type parameter as an argument and returns
+that type with an optional ``_id`` field.
```

Comment:
issue: Can this be reworded from an if clause? The action seems to be inserting without specifying an `_id`. Maybe something like:

> To insert documents of a type that requires an _id field (note: this seems like we should just tell people to do this instead of presenting it as an option, may want to ask the node team if they'd consider this a best practice), use the ``OptionalId`` type to wrap your document type.

Reply:
I'm not sure the suggested change captures the intended meaning of the existing text. From [Github issue](https://github.com/mongodb/node-mongodb-native/pull/3100#issuecomment-1010365060):

> On the other hand, on inserts, we have to guess the use case based on the schema, keeping in mind that the _id can be omitted entirely, provided explicitly, or provided indirectly via a pkFactory option. In order to accommodate users who provide the _id explicitly and definitely do want the type to be strictly enforced on inserts, we can’t make the _id always optional, which means we have to differentiate between optional and required schema _id in genera

I do not believe it is an accurate statement that all readers who require `_id` in their Schema also should use `OptionalId`. I believe the `OptionalId` use case is as follows:

> For your use case, if the collection type is something that is used independently in other parts of the code base as a document model, we recommend simply wrapping it in our OptionalId helper type when passing it at the collection creation point: db.collection<OptionalId<Test>>; alternatively, you could specify the collection interface with an optional or omitted _id property and use the WithId wrapper to access the type elsewhere, e.g., for testing.

Based off of this, I believe the "If statement" correctly captures the small subset of readers this applies to (you both muse say the `_id` is required (for accurately modeling documents) but you don't want to specify `_id` values on insert).

Let me know if my line of reasoning in this comment makes sense or if you still see issues in the text!

Reply:
Added "[as required](https://github.com/mongodb/docs-node/pull/294/commits/99efcd478a0bd42ce4609b383547eb34039cb5cc)" to the preceding text to try and clarify the use case.

Reply:
I think this needs to be asked, not inferred. The way I'm reading this, they must either specify a PKFactory, OptionalId, or provide their own _id on every insert. If I'm using the driver, I'd want to go with the "least" work option possible. This reads "Do this medium work, or do a lot of work". That seems wrong, hence why I'd like clarification from the Node team.

Reply:
Discussed over slack. Agreed on moving note lower, and making a series of follow up tickets to break the TS page into a drawer.